### PR TITLE
Add unlocking logic

### PR DESCRIPTION
**Pull request by Gr33ndev**
Part of Gamify-IT/issues#225

### CODE REVIEW

**Review by cdingm — APPROVED** [review]

**Review by WuestFlorian — APPROVED** [review]

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java [hunk 1]
```diff
@@ -127,6 +136,77 @@ public PlayerStatisticDTO updatePlayerStatisticInCourse(
...
+        for (Dungeon dungeon : currentWorld.getDungeons()) {
+          if (dungeon.isActive()) {
+            playerStatistic.addUnlockedArea(dungeon);
+            return;
+          }
+        }
+
+        worldRepository
+          .findByIndexAndCourseId(currentWorld.getIndex() + 1, courseId)
+          .ifPresent(playerStatistic::addUnlockedArea);
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]: Either use a comment, or preferably its own method.

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java [hunk 2]
```diff
@@ -127,6 +136,77 @@ public PlayerStatisticDTO updatePlayerStatisticInCourse(
...
+        for (Dungeon dungeon : otherDungeons) {
+          if (dungeon.isActive()) {
+            playerStatistic.addUnlockedArea(dungeon);
+            return;
+          }
+        }
+
+        worldRepository
+          .findByIndexAndCourseId(currentDungeon.getWorld().getIndex() + 1, courseId)
+          .ifPresent(playerStatistic::addUnlockedArea);
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]: Either a comment, or preferably its own method.

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java [hunk 3]
```diff
@@ -127,6 +136,77 @@ public PlayerStatisticDTO updatePlayerStatisticInCourse(
...
+      playerStatistic.getId()
+    );
+
+    int courseId = playerStatistic.getCourse().getId();
+
+    if (isAreaCompleted(currentArea, playerTaskStatistics)) {
+      if (currentArea instanceof World currentWorld) {
+        currentWorld.getDungeons().sort(Comparator.comparingInt(Area::getIndex));
+
+        for (Dungeon dungeon : currentWorld.getDungeons()) {
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java:158]:
```suggestion
        // As we are in a world, try to unlock the first available dungeon
        for (Dungeon dungeon : currentWorld.getDungeons()) {
```

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java [hunk 4]
```diff
@@ -127,6 +136,77 @@ public PlayerStatisticDTO updatePlayerStatisticInCourse(
...
+          if (dungeon.isActive()) {
+            playerStatistic.addUnlockedArea(dungeon);
+            return;
+          }
+        }
+
+        worldRepository
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java:166]:
```suggestion
        // No dungeon was available - try to unlock the next world
        worldRepository
```

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java (hunk 2 above)
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]:
```suggestion
        // First try to unlock the next available dungeon in the world, or if that fails try to unlock the next world
        currentDungeon
          .getWorld()
          .getDungeons()
          .stream()
          .filter(Dungeon::isActive)
          .filter(dungeon -> dungeon.getIndex() > currentDungeon.getIndex())
          .sorted(Comparator.comparingInt(Dungeon::getIndex))
          .findFirst()
          .ifPresentOrElse(playerStatistic::addUnlockedArea, () -> worldRepository
          .findByIndexAndCourseId(currentDungeon.getWorld().getIndex() + 1, courseId)
          .ifPresent(playerStatistic::addUnlockedArea))
```

File: .github/workflows/deploy-docker-image-on-release.yml [hunk 5]
```diff
@@ -4,6 +4,7 @@ on:
   push:
     branches:
       - "main"
+      - "feature/add-unlocking-logic"
```
Comment (delvh) [inline .github/workflows/deploy-docker-image-on-release.yml]:
```suggestion
```

**Pull request merged**